hkl/tests/test_save_restore_UB.py

```diff
@@ -0,0 +1,250 @@
+    assert xarray_data.fourc_geometry_name == "E4CV"
+    assert xarray_data.fourc_diffractometer_name == "fourc"
+    assert xarray_data.fourc_sample_name == "Si"
+    numpy.testing.assert_array_equal(xarray_data.fourc__pseudos, ["h k l".split()])
+    numpy.testing.assert_array_equal(xarray_data.fourc__reals, ["omega chi phi tth".split()])
+
+
+def test_fourc_run_orientation_info(fourc):
+    cat = databroker.temp().v2
+    RE = RunEngine({})
+    RE.subscribe(cat.v1.insert)
+
+    RE(bp.count([fourc]))
+    info = hkl.util.run_orientation_info(cat[1])
+    assert info is not None
+    assert isinstance(info, dict)
+    assert "fourc" in info
+    fourc_orient = info["fourc"]
+    assert "orientation_attrs" in fourc_orient
+    assert "reflections_details" in fourc_orient["orientation_attrs"]
+    assert "reflections_details" in fourc_orient
+    refls = fourc_orient["reflections_details"]
+    assert len(refls) == 2
+    assert "wavelength" in refls[0]
+    assert refls[0]["wavelength"] == 1.54
+
+
+def test_list_orientation_runs(fourc, kappa):
+    cat = databroker.temp().v2
+    RE = RunEngine({})
+    RE.subscribe(cat.v1.insert)
+    det = hw().noisy_det
+
+    def scans():
+        yield from bp.count([det])
+        yield from bp.count([fourc])
+        yield from bp.count([kappa])
+        yield from bp.count([fourc, kappa])
+
+    RE(scans())
+    runs = hkl.util.list_orientation_runs(cat)
+    # four sets of orientation info
+    # (last scan has 2, first scan has none)
+    assert len(runs.scan_id) == 4
+    assert 1 not in runs.scan_id.to_list()  # no orientation
+    assert runs.scan_id.to_list() == [2, 3, 4, 4]
+    assert runs.diffractometer_name.to_list() == "fourc kappa fourc kappa".split()
+
+
+def test_restore_orientation(fourc):
+    cat = databroker.temp().v2
+    RE = RunEngine({})
+    RE.subscribe(cat.v1.insert)
+
+    RE(bp.count([fourc]))
+    fourc_orient = hkl.util.run_orientation_info(cat[-1])["fourc"]
+
+    e4cv = Fourc("", name="e4cv")
+    assert len(e4cv.calc._samples) == 1
+
+    # typical case : good match, restores successfully
+    hkl.util.restore_orientation(fourc_orient, e4cv)
+    assert len(e4cv.calc._samples) == 2
+    sample = e4cv.calc.sample
+    assert sample.name == "Si"
+    assert len(sample.reflections) == 2
+    numpy.testing.assert_array_equal(sample.reflections, [[4, 0, 0], [0, 4, 0]])
+    # fmt: off
+    numpy.testing.assert_array_almost_equal(
+        fourc.calc.sample.UB,
+        e4cv.calc.sample.UB,
+    )
+    # fmt: on
+
+    # geometry mismatch, cannot restore
+    k4cv = Kappa("", name="k4cv")
+    assert len(k4cv.calc._samples) == 1
+    assert len(k4cv.calc.sample.reflections) == 0
+    with pytest.raises(ValueError) as exinfo:
+        hkl.util.restore_orientation(fourc_orient, k4cv)
+    args = exinfo.value.args
+    assert len(args) == 1
+    expected = "Geometries do not match: " "Orientation=E4CV, k4cv=K4CV, will not restore."
```

Comment:
Looks like `black` auto-formatting.
```suggestion
    expected = "Geometries do not match: Orientation=E4CV, k4cv=K4CV, will not restore."
```